Extend Transformers Trainer Class to Enable PyTorch Torchscript for Inference

jianan-gu opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

This PR intends to extend Transformers Trainer Class with PyTorch [Torchscript](https://pytorch.org/docs/stable/generated/torch.jit.trace.html) (`torch.jit.trace`) to speed up model inference with just-in-time compilation.

Users could simply enable "--jit_mode" Trainer input args to get benifits.


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

jianan-gu committed:
add jit mode option and model wrap

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_17153). All of your documentation changes will be reflected on that endpoint.

jianan-gu committed:
Merge branch 'main' into Introduce_Jit

lewtun commented:
> This may be more [optimum](https://github.com/huggingface/optimum/)'s domain than Transformers, what do you think @mfuntowicz @LysandreJik @lewtun ?

Thanks for the ping and thank you @jianan-gu for opening the PR! 

I agree that inference-based improvements are better suited in `optimum`, so in this case we'd need to implement a dedicated `torchscript` subpackage that effectively extends the `Trainer` in this manner. 

Having said that, I wonder if it's overkill to create a new package just to support this feature, which is natively included in PyTorch. 

@mfuntowicz @michaelbenayoun do you see any other benefits that could be had by having a dedicated subpackage for `torchscript` in `optimum`?

stas00 commented:
I'd be happy to work on this PR if you give me a green light that you'll accept this feature.

sgugger commented on src/transformers/trainer.py:
```diff
@@ -1096,7 +1096,35 @@ def call_model_init(self, trial=None):
+        if not training:
+            jit_inputs = ()
+            for _, batch in enumerate(dataloader):
+                for _, label in enumerate(batch):
```
I'm not sure I understand what is happening here. The label name is not necessarily the last key in the batch, there is a `self.label_names` attribute for this.

Also why use `enumerate` if it's to discard the index?

jianan-gu replied:
Here what we do is to construct the `jit_inputs `for `torch.jit.trace`, as steps:
(1) get the first batch from the dataloader as the `example_batch`
(2) enumerate each tensor from this `example_batch` and create each `example_tensor`
(3) tuple the `jit_inputs ` and it will be used in `torch.jit.trace`.

Therefore we do not need `self.label_names`, we are enumerating the datasets tensors ( batch: {key : tensor}, not related to labels)

Thanks.

sgugger commented on src/transformers/trainer.py:
```diff
@@ -1096,7 +1096,35 @@ def call_model_init(self, trial=None):
+            for _, batch in enumerate(dataloader):
+                for _, label in enumerate(batch):
+                    dumpy_tensor = torch.ones_like(batch[label])
+                    L1 = list(jit_inputs)
+                    L1.append(dumpy_tensor)
```
Please use a better variable name that is more descriptive of what `L1` is.

jianan-gu replied:
`L1 `was a temp list, here made changes to just use `jit_inputs` list  to avoid confusing, thanks.

sgugger commented on src/transformers/trainer.py:
```diff
@@ -1096,7 +1096,35 @@ def call_model_init(self, trial=None):
+                break
+            try:
+                if self.args.bf16:
+                    with self.autocast_smart_context_manager(), torch.no_grad():
+                        model = torch.jit.trace(model, jit_inputs, strict=False)
+                    model = torch.jit.freeze(model)
+                else:
+                    with torch.no_grad():
+                        model = torch.jit.trace(model, jit_inputs, strict=False)
+                    model = torch.jit.freeze(model)
```
Let's group all of this in one line without test. We can build the proper list of context managers (`[torch.no_grad()]` and then append to it the other one if needed) and wrap it in `ContextManagers` (that you can find in utils).

jianan-gu replied:
Have made changes to refine the code accordingly, thanks.

sgugger commented on src/transformers/training_args.py:
```diff
@@ -234,6 +234,8 @@ class TrainingArguments:
             same seed as `seed`. This can be used to ensure reproducibility of data sampling, independent of the model
             seed.
+        jit_mode(`bool`, *optional*, defaults to `False`):
+            Try to use PyTorch jit trace for inference.
```
```suggestion
            Whether ot not to use PyTorch jit trace for inference.
```

sgugger commented on src/transformers/training_args.py:
```diff
@@ -583,6 +585,7 @@ class TrainingArguments:
     no_cuda: bool = field(default=False, metadata={"help": "Do not use CUDA even when it is available"})
     seed: int = field(default=42, metadata={"help": "Random seed that will be set at the beginning of training."})
     data_seed: Optional[int] = field(default=None, metadata={"help": "Random seed to be used with data samplers."})
+    jit_mode: bool = field(default=False, metadata={"help": "Try to use PyTorch jit trace for inference"})
```
```suggestion
    jit_mode: bool = field(default=False, metadata={"help": "Whether or not to use PyTorch jit trace for inference"})
```

sgugger reviewed (commented):
Thanks for adding this. We would need some test before we can merge the PR, and the code needs to be adapted to the way the labels are handled in Transformers.

jianan-gu committed:
Update src/transformers/training_args.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

jianan-gu committed:
Update src/transformers/training_args.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

jianan-gu committed:
Merge branch 'huggingface:main' into Introduce_Jit

jianan-gu committed:
refine code

sgugger commented on src/transformers/trainer.py:
```diff
@@ -1140,7 +1141,28 @@ def call_model_init(self, trial=None):
+    def torch_jit_model(self, model, training=False, dataloader=None):
+        if not training:
+            jit_inputs = []
+            example_batch = list(dataloader)[0]
```
No, we can't unpack the whole dataloader as it might be very slow and make the user run out of RAM if the dataset is huge.
```suggestion
            for example_batch in dataloader:
                break
```

sgugger commented on src/transformers/trainer.py:
```diff
@@ -1140,7 +1141,28 @@ def call_model_init(self, trial=None):
+        if not training:
+            jit_inputs = []
+            example_batch = list(dataloader)[0]
+            for _, key in enumerate(example_batch):
```
```suggestion
            for key in example_batch:
```
Again, why enumerate when the index is not used?

sgugger reviewed (commented):
Thanks for addressing the comments, this will also need a test before we can merge.

jianan-gu committed:
Update src/transformers/trainer.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

jianan-gu committed:
Update src/transformers/trainer.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

jianan-gu committed:
add ut and refine code

jianan-gu committed:
Merge branch 'main' into Introduce_Jit

jianan-gu committed:
code refine

jianan-gu commented:
Adding UT Tests in test_trainer.py for jit which covers `evaluate `and `predict`.
Thanks.

sgugger commented on src/transformers/trainer.py:
```diff
@@ -1158,6 +1158,27 @@ def call_model_init(self, trial=None):
+                jit_model(**example_batch)
+                model = jit_model
+            except (RuntimeError, TypeError):
+                logger.warning("fail to use PyTorch jit mode")
```
Maybe include the error message in the warning here?

jianan-gu replied:
have made changes accordingly, thanks.

sgugger commented on src/transformers/training_args.py:
```diff
@@ -236,6 +236,8 @@ class TrainingArguments:
             Random seed to be used with data samplers. If not set, random generators for data sampling will use the
             same seed as `seed`. This can be used to ensure reproducibility of data sampling, independent of the model
             seed.
+        jit_mode(`bool`, *optional*, defaults to `False`):
```
We should maybe add an `eval` suffix here to make it clear to the user this is inference only.

jianan-gu replied:
sure, have made changes, thanks for the suggestion.

sgugger reviewed (approved):
Thanks for adding the tests! I have a few last small comments.

Do you want to have a second look @stas00 ?

jianan-gu committed:
refine code

stas00 commented on src/transformers/trainer.py:
```diff
@@ -1158,6 +1158,27 @@ def call_model_init(self, trial=None):
+    def torch_jit_model(self, model, training=False, dataloader=None):
+        if not training:
+            jit_inputs = []
+            for example_batch in dataloader:
+                break
```
```suggestion
            example_batch = next(iter(dataloader))
```

perhaps, otherwise it reads a bit as a beginner python code.

but also `dataloader` can be `None` , so this code needs to be more defensive.

stas00 commented on src/transformers/trainer.py:
```diff
@@ -1158,6 +1158,27 @@ def call_model_init(self, trial=None):
+                jit_model = torch.jit.freeze(jit_model)
+                jit_model(**example_batch)
+                model = jit_model
+            except (RuntimeError, TypeError) as ex:
+                logger.warning(f"fail to use PyTorch jit mode due to: {ex}.")
```
```suggestion
            except (RuntimeError, TypeError) as exception:
                logger.warning(f"failed to use PyTorch jit mode due to: {exception}.")
```

stas00 reviewed (changes requested):
What is the sounds of a tree falling in the forest when there is nobody to hear it?

All these new features need to be documented on the user-side so that they actually get used. So the same as with IPEX let's add user docs and as before ideally a small benchmark to give users an incentive to try the feature.

updade: I added a doc for inference as we haven't created it and then you can push the usage examples there - like we did with IPEX PR. So please fill out the blank and then we are good to go I think.

and added a few nits in the code.

stas00 committed:
add inference doc

jianan-gu committed:
Update src/transformers/trainer.py

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

jianan-gu committed:
Update src/transformers/trainer.py

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

jianan-gu committed:
add cpu inference performance doc

jianan-gu committed:
Update perf_infer_cpu.mdx

jianan-gu committed:
Update perf_infer_cpu.mdx

jianan-gu committed:
Update performance.mdx

jianan-gu commented on docs/source/en/perf_infer_gpu_one.mdx:
```diff
@@ -0,0 +1,18 @@
+
+## JIT-mode
+
+XXX: fill in the details
```
Yes, it is for inference only.

stas00 replied:
so why `torch_jit_model` name? let's name it accordingly?

jianan-gu replied:
Sure, have made changes accordingly, thanks for your suggestion.

stas00 replied:
Thank you, @jianan-gu - please ping me when it's ready for a final review.

jianan-gu committed:
Update _toctree.yml

jianan-gu committed:
refine jit func naming

jianan-gu committed:
Update _toctree.yml

jianan-gu committed:
Delete perf_infer_gpu_one.mdx

jianan-gu commented:
> What is the sounds of a tree falling in the forest when there is nobody to hear it?
> 
> All these new features need to be documented on the user-side so that they actually get used. So the same as with IPEX let's add user docs and as before ideally a small benchmark to give users an incentive to try the feature.
> 
> updade: I added a doc for inference as we haven't created it and then you can push the usage examples there - like we did with IPEX PR. So please fill out the blank and then we are good to go I think.
> 
> and added a few nits in the code.

Though this jit mode works both for CPU and GPU, the current IPEX release covers CPU side optimizations with jit mode for model inference. Therefore we only have added and updated the inference doc for CPU (perf_infer_cpu.mdx), please take a review of the contents @stas00 , thanks. (for the small benchmark, for now, we only have relative performance numbers like shown in https://github.com/huggingface/transformers/issues/17137, so we will prepare for the numbers with --skip_memory_metrics 0 as a follow-up)

jianan-gu committed:
Update perf_infer_cpu.mdx

stas00 commented on docs/source/en/perf_infer_cpu.mdx:
```diff
@@ -0,0 +1,57 @@
+For a gentle introduction to TorchScript, see the Introduction to [PyTorch TorchScript tutorial](https://pytorch.org/tutorials/beginner/Intro_to_TorchScript_tutorial.html#tracing-modules).
+
+### IPEX Graph Optimization with JIT-mode
+Intel® Extension for PyTorch provides further optimizations in jit mode for Transformers series models. It is highly recommended for users to take advantage of Intel® Extension for PyTorch with jit mode. Some frequently used operator patterns from Transformers models are already supported in Intel® Extension for PyTorch with jit mode fusions. Those fusion patterns like Multi-head-attention fusion ...
```
```suggestion
Intel® Extension for PyTorch provides further optimizations in jit mode for Transformers series models. It is highly recommended for users to take advantage of Intel® Extension for PyTorch with jit mode. Some frequently used operator patterns from Transformers models are already supported in Intel® Extension for PyTorch with jit mode fusions. Those fusion patterns like Multi-head-attention fusion, Concat Linear, Linear+Add, Linear+Gelu, Add+LayerNorm fusion and etc. are enabled and perform well. The benefit of the fusion is delivered to users in a transparent fashion. According to the analysis, ~70% of most popular NLP tasks in question-answering, text-classification, and token-classification can get performance benefits with these fusion patterns for both Float32 precision and BFloat16 Mixed precision.
```

stas00 commented on src/transformers/trainer.py:
```diff
@@ -1158,6 +1158,26 @@ def call_model_init(self, trial=None):
 
         return model
 
+    def torch_jit_model_eval(self, model, training=False, dataloader=None):
+        if not training:
+            jit_inputs = []
+            example_batch = next(iter(dataloader))
```
is `dataloader` required? As I commented before you're not checking anywhere if it's None. 

If it's required then please remove `=None` in the function signature.

jianan-gu replied:
yes, it is required for jit.
Thanks for your suggestion, here have made changes to add a checking before constructing the jit inputs and removed `=None` in the function signature.

stas00 replied:
All looks good now to me.

Thank you!

stas00 reviewed (changes requested):
last few small things to adjust and then it should be good.

jianan-gu committed:
Update docs/source/en/perf_infer_cpu.mdx

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

jianan-gu committed:
add none check before jit

stas00 commented:
@sgugger, would you like to have a quick look at the last changes since your review - mainly the newly added doc. Thank you!

otherwise it's good to be merged.

sgugger commented on docs/source/en/perf_infer_cpu.mdx:
```diff
@@ -0,0 +1,57 @@
+Take an example of the use cases on [Transformers question-answering](https://github.com/huggingface/transformers/tree/main/examples/pytorch/question-answering)
+
+- Inference using jit mode on CPU:
+<pre> python run_qa.py \
+--model_name_or_path csarron/bert-base-uncased-squad-v1 \
+--dataset_name squad \
+--do_eval \
+--max_seq_length 384 \
+--doc_stride 128 \
+--output_dir /tmp/ \
+--no_cuda \
+<b> --jit_mode_eval </b></pre> 
```
Not a big fan of using the HTML syntax here, I'd prefer the triple backticks, and it's okay if the new arg is not in bold. If keeping it, some spaces need to be removed.
```suggestion
<pre>python run_qa.py \
--model_name_or_path csarron/bert-base-uncased-squad-v1 \
--dataset_name squad \
--do_eval \
--max_seq_length 384 \
--doc_stride 128 \
--output_dir /tmp/ \
--no_cuda \
<b>--jit_mode_eval </b></pre> 
```

sgugger commented on docs/source/en/perf_infer_cpu.mdx:
```diff
@@ -0,0 +1,57 @@
+<b> --jit_mode_eval </b></pre> 
+
+- Inference with IPEX using jit mode on CPU:
+<pre> python run_qa.py \
+--model_name_or_path csarron/bert-base-uncased-squad-v1 \
+--dataset_name squad \
+--do_eval \
+--max_seq_length 384 \
+--doc_stride 128 \
+--output_dir /tmp/ \
+--no_cuda \
+<b> --use_ipex \ </b>
+<b> --jit_mode_eval </b></pre> 
```
Same comment
```suggestion
<pre>python run_qa.py \
--model_name_or_path csarron/bert-base-uncased-squad-v1 \
--dataset_name squad \
--do_eval \
--max_seq_length 384 \
--doc_stride 128 \
--output_dir /tmp/ \
--no_cuda \
<b>--use_ipex \</b>
<b>--jit_mode_eval</b></pre> 
```

sgugger reviewed (approved):
Some small comments on the new doc, thanks for adding it!

jianan-gu committed:
Update docs/source/en/perf_infer_cpu.mdx

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

jianan-gu committed:
Update docs/source/en/perf_infer_cpu.mdx

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>
